### PR TITLE
Epochs

### LINKED ISSUE #869: [Capacity] Epoch time to claim my un-staked tokens
### Description:

Implementation of how to Replenish Replenishable means that all Capacity is replenished after a fixed period called an Epoch Period. An Epoch Period is composed of a set number of blocks. In the example below, the Epoch Period is three blocks. The initial Epoch Period will be around 100 blocks. 

Acceptance criteria :
- Must be able to query epoch number in storage
- After a fixed number of blocks, a new Epoch Period begins.
- At the start of a new Epoch Period, `EpochNumber` storage is increased by 1.
- At the start of a new block, `CurrentBlockUsedCapacity` storage is reset.
- Benchmark for `on_initialize`

### PR DESCRIPTION
**Pull request by shannonwells**
# Goal
The goal of this PR is to implement epochs getting and setting, including starting an epoch when it's time to do so.

Closes #869 

# Discussion

# Checklist
- [x] Must be able to query epoch number in storage
- [x] After a fixed number of blocks, a new Epoch Period begins.
- [x] At the start of a new Epoch Period, EpochNumber storage is increased by 1.
- [x] At the start of a new block, CurrentBlockUsedCapacity storage is reset.
- [x] Design doc(s) updated
- [x] Tests added
- [x] Benchmarks (for on_initialize) added
- [ ] Weights updated

N/A
- Chain spec updated
- Custom RPC OR Runtime API added/changed? Updated js/api-augment.

### CODE REVIEW

File: pallets/capacity/src/lib.rs [hunk 1]
```diff
@@ -47,12 +47,13 @@
...
 	ensure,
-	traits::{Currency, Get, LockIdentifier, LockableCurrency, WithdrawReasons},
+	traits::{Currency, Get, Hooks, LockIdentifier, LockableCurrency, WithdrawReasons},
+	weights::{constants::RocksDbWeight, Weight},
 };
 
 use sp_runtime::{
-	traits::{CheckedAdd, Saturating, Zero},
-	ArithmeticError, DispatchError, Perbill,
+	traits::{CheckedAdd, Zero},
```
Comment (shannonwells) [inline pallets/capacity/src/lib.rs:55]: the IDE did it.

File: pallets/capacity/src/lib.rs [hunk 2]
```diff
@@ -126,6 +127,11 @@ pub mod pallet {
 		/// The number of Epochs before you can unlock tokens after unstaking.
 		#[pallet::constant]
 		type UnstakingThawPeriod: Get<u16>;
+
+		/// Maximum number of blocks an epoch can be
+		/// currently used as the actual value of epoch length.
+		#[pallet::constant]
+		type MaxEpochLength: Get<Self::BlockNumber>;
```
Comment (shannonwells) [inline pallets/capacity/src/lib.rs:134]: don't know if we should keep this name or not, it's not _technically_ premature optimization...

File: pallets/capacity/src/testing_utils.rs [hunk 3]
```diff
@@ -1,10 +1,7 @@
 use super::*;
 use crate::mock::*;
 
-use frame_support::{
-	assert_ok,
-	traits::{OnFinalize, OnInitialize},
-};
+use frame_support::assert_ok;
```
Comment (shannonwells) [inline pallets/capacity/src/testing_utils.rs:4]: to fix some warnings.

File: pallets/capacity/src/tests.rs [hunk 4]
```diff
@@ -688,19 +687,18 @@ fn staking_account_details_decrease_by_reduces_active_staking_balance_and_create
 			total: BalanceOf::<Test>::from(10u64),
 			unlocking: BoundedVec::default(),
 		};
-		staking_account_details.decrease_by(3, 3);
+		staking_account_details.decrease_by(3, 3).expect("decrease_by failed");
```
Comment (shannonwells) [inline pallets/capacity/src/tests.rs:690]: also to fix some warnings - complaints about unused Result

File: runtime/frequency/src/lib.rs [hunk 5]
```diff
@@ -308,14 +308,13 @@ impl pallet_capacity::Config for Runtime {
...
+	type MinimumStakingAmount = CapacityMinimumStakingAmount;
 	type TargetValidator = Msa;
-	type MaxUnlockingChunks = ConstU32<4>;
-
+	type MaxUnlockingChunks = CapacityMaxUnlockingChunks;
 	#[cfg(feature = "runtime-benchmarks")]
 	type BenchmarkHelper = Msa;
-
-	type UnstakingThawPeriod = ConstU16<2>;
+	type UnstakingThawPeriod = CapacityUnstakingThawPeriod;
```
Comment (shannonwells) [inline runtime/frequency/src/lib.rs:316]: moved some more of these into the constants file.

File: pallets/capacity/src/lib.rs (hunk 2 above)
Comment (enddynayn) [reply pallets/capacity/src/lib.rs:134]: I would drop it, but we can update it in a different PR.

File: pallets/capacity/src/lib.rs [hunk 6]
```diff
@@ -158,6 +164,21 @@ pub mod pallet {
 	pub type CapacityLedger<T: Config> =
 		StorageMap<_, Twox64Concat, MessageSourceId, CapacityDetails<BalanceOf<T>, T::BlockNumber>>;
 
+	/// Storage for the current epoch number
+	#[pallet::storage]
+	#[pallet::getter(fn get_current_epoch)]
+	pub type CurrentEpoch<T: Config> = StorageValue<_, T::BlockNumber, ValueQuery>;
```
Comment (enddynayn) [inline pallets/capacity/src/lib.rs:170]: We may want to create a ticket to whitelist this so it does not cost a read when accessing it.

File: pallets/capacity/src/lib.rs [hunk 7]
```diff
@@ -158,6 +164,21 @@ pub mod pallet {
...
+
+	/// Storage for the current epoch info
+	#[pallet::storage]
+	#[pallet::getter(fn get_current_epoch_info)]
+	pub type CurrentEpochInfo<T: Config> = StorageValue<_, EpochInfo<T::BlockNumber>, ValueQuery>;
+
+	/// Storage for the current epoch number
+	#[pallet::storage]
+	#[pallet::getter(fn get_current_epoch_used_capacity)]
+	pub type CurrentEpochUsedCapacity<T: Config> = StorageValue<_, BalanceOf<T>, ValueQuery>;
```
Comment (enddynayn) [inline pallets/capacity/src/lib.rs:180]: maybe for CurrentEpochInfo and CurrentEpochUsedCapacity too.

File: pallets/capacity/src/types.rs [hunk 8]
```diff
@@ -169,3 +169,15 @@ impl<Balance: Saturating + Copy + CheckedAdd, BlockNumber> CapacityDetails<Balan
...
 	}
 }
+
+/// The type for storing details about an epoch.
+#[derive(
+	PartialEq, Eq, Clone, Default, PartialOrd, Encode, Decode, RuntimeDebug, TypeInfo, MaxEncodedLen,
+)]
+
+/// Information about the current epoch.
+/// May evolve to store other needed data such as epoch_end.
```
Comment (enddynayn) [inline pallets/capacity/src/types.rs:179]: nice

File: pallets/msa/src/weights.rs [hunk 9]
```diff
@@ -74,7 +74,7 @@ impl<T: frame_system::Config> WeightInfo for SubstrateWeight<T> {
 	// Storage: Msa PublicKeyToMsaId (r:1 w:1)
```
Comment (enddynayn) [inline pallets/msa/src/weights.rs:74]: I think we can remove all the non-capacity ones.

**Review by enddynayn — APPROVED** [review]
Super clean! 💯

File: pallets/capacity/src/lib.rs (hunk 6 above)
Comment (enddynayn) [reply pallets/capacity/src/lib.rs:170]: I think we can also create a different PR for updating the fields where we are using block-number when we mean epoch number.

Comment (enddynayn) [inline pallets/capacity/src/lib.rs:170]: Let's avoid using  T::BlockNumber here and create our epoch type. We should not be able to add an epoch with a block number. Not a blocker for this PR.

File: pallets/capacity/src/lib.rs [hunk 10]
```diff
@@ -474,4 +502,25 @@ impl<T: Config> Pallet<T> {
...
+	/// Get current epoch length.
+	fn get_epoch_length() -> T::BlockNumber {
+		<T>::MaxEpochLength::get()
+	}
+
+	fn start_new_epoch_if_needed(current: T::BlockNumber) -> Weight {
+		if Self::get_current_epoch_info()
+			.epoch_start
+			.saturating_add(Self::get_epoch_length())
+			.eq(&current)
```
Comment (enddynayn) [inline pallets/capacity/src/lib.rs:515]: fancy

File: pallets/capacity/src/lib.rs (hunk 7 above)
Comment (shannonwells) [reply pallets/capacity/src/lib.rs:180]: ah you mean whitelist this storage too.

**Pull request merged**